File: lib/logfilt.ex
defmodule Logfilt do

  @match_table [
          {~r/INFO/, :green},
          {~r/WARNING/, :yellow},
          {~r/ERROR/, :red},
          {~r/CRITICAL/, :magenta},
          {~r/^.*!!!.*$/, :cyan},
          {~r/\d{4}-\d{2}-\d{2}[^]]*]/, :grey},
      ]

  @delete_lines [
    ~r/recording\.py/,
    ~r{GET /gae_mini_profiler},
    ~r/Stripped prohibited headers from URLFetch request: \['Host'\]/,
    ~r/Mismatch between XSRF header \(None\) and cookie/
  ]

  @delete_expressions [
    ~r/\d{4}-\d{2}-\d{2}[^]]*]/
  ]

  @color_map %{
                 red: "\x1b[31m",
                 green: "\x1b[32m",
                 yellow: "\x1b[33m",
                 magenta: "\x1b[35m",
                 cyan: "\x1b[36m",
                 grey: "\x1b[30m",
                 clear: "\x1b[0m",
             }


  @doc """
  Get color annotations for matches of a pattern in the provided string.

  A priority parameter controls how conflicting colorations are resolved.

  Returns: {priority, start_position, len, color}
  """
  def maybe_colorize_partial priority, color, pat, line do
    Regex.scan(pat, line, return: :index)
    |> Enum.map fn [{start, len}|_] -> {priority, start, len, color} end
  end

  @doc """
  Get the color from the color annotation prior to the specified position.

  Default to :clear if there's no annotation before the current position.
  """
  def prev_color(accumlated_colors, pos) do
    accumlated_colors
    |> Enum.sort
    |> Enum.take_while(fn {i, _} -> i <= pos end)
    |> List.last
    |> case do
         nil -> :clear
         {_, color} -> color
       end
  end

  @doc """
  Turn a list of colorization tuples into a list of positions where the color changes.

  The colorization tuples are {priority, start, len, color}.  Colorizations
  are applied according to start poistion, using the priority parameter to
  break any ties.

  Return a list of the format [{pos, color}...]
  """
  def prioritized_colorization_to_color_change_offsets colorization do
    colorization
    |> Enum.sort_by(fn {priority, start, len, color} ->
                      {start, -1 * priority, len, color} end)
    |> Enum.reduce([], fn {_, start, len, color}, acc ->
                     acc ++ [{start, color},
                             {start + len, prev_color(acc, start + len)}]
                   end)
  end

  @doc """
  Apply a list of positions where the color changes to a string.

  This will insert the appropriate escape codes into the string at those
  positions.
  """
  def apply_prioritized_colorization colorization, line do
    color_tup = (prioritized_colorization_to_color_change_offsets(colorization)
      |> Enum.sort
      |> Enum.reduce({0, ""}, fn {pos, color}, {startpos, str} ->
                       newstr = (str <> String.slice(line, startpos, pos - startpos)
                                 <> @color_map[color])
                       newpos = pos
                       {newpos, newstr}
                     end))

    {last_pos, color_str} = color_tup
    color_str <> String.slice(line, last_pos, String.length(line)) <> @color_map[:clear]
  end

  @doc """
  Colorize a string of text accorting to the rules in @match_table.
  """
  def colorize_line line do
    @match_table
    |> Enum.with_index
    |> Enum.map(fn {{pat, color}, i} -> maybe_colorize_partial(i, color, pat, line) end)
    |> List.flatten
    |> apply_prioritized_colorization(line)
  end

  def delete_expressions line do
    Enum.reduce(@delete_expressions, line, fn regex, repl -> Regex.replace(regex, repl, "") end)
  end

  @doc """
  Continually read from stdin, colorize the lines, and write to stdout.
  """
  def main(_args) do
    IO.stream(:stdio, :line)
    |> Stream.filter(fn line -> Enum.all?(@delete_lines, fn regex -> not Regex.match?(regex, line) end) end)
    |> Stream.map(&String.strip/1)
    |> Stream.map(&delete_expressions/1)
    |> Stream.map(&colorize_line/1)
    |> Enum.each(&IO.puts(&1))
  end

end

File: logfilt.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"

	"github.com/fatih/color"
)

type colorNode struct {
	c        *color.Color
	text     string
	children *[]*colorNode
}

var deleteRegexes = []*regexp.Regexp{
	regexp.MustCompile("recording\\.py"),
	regexp.MustCompile("GET /gae_mini_profiler"),
	regexp.MustCompile("Stripped prohibited headers from URLFetch request: \\['Host'\\]"),
	regexp.MustCompile("Mismatch between XSRF header \\(None\\) and cookie"),
}

var expressionsToDelete = []*regexp.Regexp{
	regexp.MustCompile("\\d{4}-\\d{2}-\\d{2}\\s*(\\d{2}:\\d{2}:\\d{2})[^]]*]"),
}

var colorRegexes = map[*regexp.Regexp](*color.Color){
	regexp.MustCompile("^.*!!!.*$"):                   color.New(color.FgCyan),
	regexp.MustCompile("^.*---.*$"):                   color.New(color.FgCyan),
	regexp.MustCompile("\\d{4}-\\d{2}-\\d{2}[^]]*]/"): color.New(color.FgHiBlack),
	regexp.MustCompile("INFO"):                        color.New(color.FgGreen),
	regexp.MustCompile("WARNING"):                     color.New(color.FgYellow),
	regexp.MustCompile("ERROR"):                       color.New(color.FgRed),
	regexp.MustCompile("CRITICAL"):                    color.New(color.FgMagenta),
}

func shouldDeleteLine(line string) bool {
	for _, re := range deleteRegexes {
		if re.MatchString(line) {
			return true
		}
	}
	return false
}

func deleteExpressions(line string) string {
	for _, re := range expressionsToDelete {
		line = re.ReplaceAllString(line, "$1")
	}
	return line
}

func colorize(node *colorNode) *colorNode {
	if node.children != nil {
		for _, child := range *node.children {
			colorize(child)
		}
		return node
	}

	for key, value := range colorRegexes {
		if key.MatchString(node.text) {
			parts := key.Split(node.text, 2)
			matchText := key.FindString(node.text)
			if parts[0] != "" || parts[1] != "" {
				children := []*colorNode{
					&colorNode{c: node.c, text: parts[0], children: nil},
					&colorNode{c: value, text: matchText, children: nil},
					&colorNode{c: node.c, text: parts[1], children: nil},
				}
				node.children = &children
				return colorize(node)
			}
			node.c = value
		}
	}
	return node
}

func (node colorNode) toString() string {
	if node.children == nil {
		return node.c.Sprint(node.text)
	}

	result := ""
	for _, child := range *node.children {
		result += child.toString()
	}
	return result
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		line := scanner.Text()
		if shouldDeleteLine(line) {
			continue
		}
		line = deleteExpressions(line)
		colorExpr := colorNode{
			c:        color.New(color.Reset),
			text:     line,
			children: nil,
		}
		colorized := colorize(&colorExpr)
		fmt.Println(colorized.toString())
	}
}
